### ISSUE TITLE
Docs missing a full list of node types and what they mean

### ISSUE THREAD

**Issue by Fishrock123** [eslint/eslint#11065]
A full & descriptive list of node types seems to be missing from the docs. For example, I can't find `ExpressionStatement` anywhere. This is really limiting my ability to build new / modify existing eslint rules.

**Edit**: I also see that the expressions can have a further `type`, such as `'CallExpression'`. I also cannot find much on those.

**Edit 2**: I just found https://github.com/eslint/eslint/blob/caeb223c4f7b0b6fe35e5348ae0df4c6446b5bed/docs/developer-guide/code-path-analysis.md but strangely it is missing from the website?) _(Also it still doesn't have this.)_

-----

**The version of ESLint you are using.**
latest

**The problem you want to solve.**
I want to know what `ExpressionStatement` is (in detail) right now, but really, there should be a list of all node types.

**Your take on the correct solution to problem.**
Document this

**Are you willing to submit a pull request to implement this change?**
Probably not, someone who knows what these mean should.

**Comment by not-an-aardvark** [comment]
Hi, thanks for creating an issue.

ASTs from ESLint follow the [ESTree](https://github.com/estree/estree) specification, and there is a list of node types at that repo. Are those similar to what you're looking for?

An `ExpressionStatement` is the term for a *statement* in source code which consists solely of an *expression* (in contrast to other types of statements, like an `IfStatement` (containing an `if`/`else` block, or a `WhileStatement`, which contains a while loop.

For example, the following function body starts with two `ExpressionStatements` and has an `IfStatement` after it:

```js
function foo() {
    // An `ExpressionStatement` whose expression happens to be a `CallExpression`.
    baz();

    // An `ExpressionStatement` whose expression happens to be a number literal.
    // (This is useless, but valid JS.)
    5;
  
    // An `IfStatement`
    if (x) y();
}
```

You might also be interested in [ASTExplorer](https://astexplorer.net/) for viewing what the AST looks like for different pieces of code.

**Comment by Fishrock123** [comment]
That seems like what I'm looking for, yes. I really wish some version of this was on the ESLint website.

As for ASTExplorer - I had found it, but forgot to look for the actual AST parser (espree) that ESLint uses, so I didn't think I could rely on the output shapes.

**Comment by not-an-aardvark** [comment]
Thanks, this is good docs feedback. FWIW it is linked from [here](https://eslint.org/docs/developer-guide/working-with-custom-parsers#the-ast-specification) and briefly mentioned [here](https://eslint.org/docs/developer-guide/working-with-rules#rule-basics) under the description of the `create` method, but it definitely seems easy to miss. We could probably update the docs to make those references more visible.

**Comment by kaicataldo** [comment]
@Fishrock123 Any interest in making a PR since you have some first-hand experience and might have more insight into what would be helpful for end users?
